src/proxyp/Message.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_PROXYP_MESSAGE_H
+#define SQUID_PROXYP_MESSAGE_H
+
+#include "base/RefCount.h"
+#include "ip/Address.h"
+#include "proxyp/Protocol.h"
+#include "sbuf/SBuf.h"
+
+namespace ProxyProtocol {
+
+/// PROXY protocol v1 or v2 message
+class Message: public RefCountable
+{
+public:
+    typedef RefCount<Message> Pointer;
+    typedef std::vector<Two::Tlv> Tlvs;
+    typedef std::map<SBuf, Two::HeaderType> FieldMap;
+
+    Message(const char *ver, const uint8_t cmd = Two::cmdProxy);
+
+    /// HTTP header-like string representation of the message.
+    /// The returned string has several mandatory lines for the protocol
+    /// version, command addresses and ports:
+    /// :version: version CRLF
+    /// :command: command CRLF
+    /// :src_addr: srcAddr CRLF
+    /// :dst_addr: dstAddr CRLF
+    /// :src_port: srcPort CRLF
+    /// :dst_port: dstPort CRLF
+    /// and may also contain several optional lines for each TLV:
+    /// type: value CRLF
+    SBuf getAll(const char sep) const;
+
+    /// \returns a delimiter-separated list of values of TLVs of the given type
+    SBuf getValues(const uint32_t headerType, const char delimiter = ',') const;
+
+    /// Searches for the first key-value pair occurrence within the
+    /// value for the provided TLV type. Assumes that the TLV value
+    /// is a list of delimiter-separated items and the items are
+    /// pairs separated by '='.
+    /// \returns the value of the found pair or an empty string.
+    SBuf getElem(const uint32_t headerType, const char *member, const char delimiter) const;
+
+    /// the message version
+    const char *version() const { return version_; }
+
+    /// whether source and destination addresses are valid addresses of the original "client" connection
+    bool hasForwardedAddresses() const { return !localConnection() && hasAddresses(); }
+
+    /// marks the message as lacking address information
+    void ignoreAddresses() { ignoreAddresses_ = true; }
+
+    /// whether the message relays address information (including LOCAL connections)
+    bool hasAddresses() const { return !ignoreAddresses_; }
+
+    /// whether both source and destination addresses have tcpVersion
+    bool hasMatchingTcpVersion(const SBuf &tcpVersion) const;
+
+    /// a mapping between pseudo header names and ids
+    static FieldMap PseudoHeaderFields;
```

Comment:
I think this belongs to the ProxyProtocol namespace, not the Message class. You can place this object into `Elements.*`.